### PR TITLE
Minor: Improve documentation about `OnceAsync`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes #.

## Rationale for this change
@comphead  said this would be amazing :) : https://github.com/apache/datafusion/pull/13203#discussion_r1825932243

> Would be amazing to clarify what is OnceAsync and why clone is not fit for it?

So I tried to improve the documentation 

## What changes are included in this PR?

Improve documentation

## Are these changes tested?
N/A 

## Are there any user-facing changes?

No, just comments

### CODE REVIEW

**Review by comphead — APPROVED** [review]
Thanks @alamb that was pretty interesting to read. So this is sort of memory barrier where multiple consumers can hold the resource but its released whenever first consumer send OnceAsync?

**Comment by alamb** [comment]
> Thanks @alamb that was pretty interesting to read. So this is sort of memory barrier where multiple consumers can hold the resource but its released whenever first consumer send OnceAsync?

I think it is more like the shared rendezvous point for all the output streams -- aka to start producing data out of each partition, the left side is read in entirely. The way the output streams wait for the left side to be read is by waiting on the `OnceAsync` -- and once it resolves all the output streams get a (read only) copy of the data to work on

**Comment by alamb** [comment]
Perhaps @korowa  has some additional suggestions on how to improve the wording or something

File: datafusion/physical-plan/src/joins/hash_join.rs [hunk 1]
```diff
@@ -314,6 +316,11 @@ pub struct HashJoinExec {
     /// if there is a projection, the schema isn't the same as the output schema.
     join_schema: SchemaRef,
     /// Future that consumes left input and builds the hash table
+    ///
+    /// This structure is *shared* across all output streams.
```
Comment (korowa) [inline datafusion/physical-plan/src/joins/hash_join.rs:320]:
```suggestion
    /// For CollectLeft partition mode, this structure is *shared* across all output streams.
```

File: datafusion/physical-plan/src/joins/cross_join.rs [hunk 2]
```diff
@@ -46,16 +46,23 @@ use datafusion_physical_expr::equivalence::join_equivalence_properties;
...
+///
+/// Buffers the left input into memory and then streams batches from each
+/// partition on the right input combining them with the buffered left input
+/// to generate the output.
+///
+/// # Clone / Shared State
+///
+/// Note this structure includes a [`OnceAsync`] that is used to coordinate the
+/// loading of the left side with the processing in each output stream.
+/// Therefore it can not be [`Clone`]
```
Comment (korowa) [inline datafusion/physical-plan/src/joins/cross_join.rs:65]: Disregarding this PR -- I wonder is there any difference between cloning `OnceAsync` and `Arc<dyn ExecutionPlan>` fields?

It just seems that cloning both types is error prone for the cases when original operator is not dropped after being clones, and both clone and original trying to execute -- if so, and there is no significant difference between these types, maybe we should make `OnceAsync` cloneable, or forbid to clone almost all `PhysicalPlan`s?

Comment (alamb) [reply datafusion/physical-plan/src/joins/cross_join.rs:65]: I think calling `ExecutionPlan::execute` results in a `Stream` 

In the join cases, the data would be shared across the clone, I think

For example, 

The first call to `HashJoinExec::execute`  (perhaps non obviously) updates the contents of `self.left_fut`  

https://github.com/apache/datafusion/blob/a32802c69cfb74ff5750700a8b72db87a44c55f0/datafusion/physical-plan/src/joins/hash_join.rs#L713-L712

And then subsequent invocations of HashJoinExec::exec read the (same) `self.left_fut` which is shared across streams

Comment (korowa) [reply datafusion/physical-plan/src/joins/cross_join.rs:74]: Yes, this is correct.

**Review by korowa — APPROVED** [review]
LGTM, thank you @alamb

**Pull request merged**